src/typescript-worker/typescript.ts

```diff
@@ -145,19 +154,82 @@ export class TypeScriptBuilder {
   }
 }
 
+interface VersionedFile {
+  version: number;
+  content: string;
+}
+
 class WorkerLanguageServiceHost implements ts.LanguageServiceHost {
   readonly compilerOptions: ts.CompilerOptions;
   readonly packageRoot: string;
-  readonly files: Map<string, string>;
+  readonly files: Map<string, VersionedFile>;
 
   constructor(
-    files: Map<string, string>,
     packageRoot: string,
-    compilerOptions: ts.CompilerOptions
+    compilerOptions: ts.CompilerOptions,
+    files?: Map<string, VersionedFile>
   ) {
     this.packageRoot = packageRoot;
     this.compilerOptions = compilerOptions;
-    this.files = files;
+    this.files = files || new Map();
+  }
+
+  /*
+   *  When a new new "process" command is received, we iterate through all of the files,
+   *  and update files accordingly depending on if they have new content or not.
+   *
+   *  With how the TS API works, we can use simple versioning to tell the
+   *  Language service that a file has been updated
+   *
+   *  If the file submitted is a new file, we add it to our collection
+   */
+  updateFileContentIfNeeded(fileName: string, content: string) {
+    const file = this.files.get(fileName);
+    if (file && this._fileContentHasUpdated(file, content)) {
+      file.content = content;
+      file.version += 1;
+    } else {
+      this.addFile(fileName, {content, version: 0});
+    }
+  }
+
+  /**
+   * Sync up the freshly acquired project files.
+   * In the syncing process files yet to be added are added, and versioned.
+   * Files that existed already but are modified are updated, and their version number
+   * gets bumped fo that the languageservice knows to update these files.
+   * */
+  sync(files: Map<string, string>) {
+    files.forEach((file, fileName) =>
+      this.updateFileContentIfNeeded(fileName, file)
+    );
+    this.removeDeletedFiles(files);
+  }
+
+  removeDeletedFiles(files: Map<string, string>) {
+    this.getScriptFileNames().forEach((fileName) => {
+      if (!files.has(fileName)) {
+        this.removeFile(fileName);
+      }
+    });
+  }
+
+  addFile(fileName: string, file: VersionedFile) {
+    this.files.set(fileName, file);
+  }
+
+  addFiles(files: Map<string, VersionedFile>) {
+    files.forEach((file, fileName) => {
+      this.files.set(fileName, file);
+    });
+  }
+
+  removeFile(fileName: string) {
+    this.files.delete(fileName);
+  }
+
+  _fileContentHasUpdated(oldFile: VersionedFile, newContent: string) {
```

Comment:
```suggestion
  private _fileContentHasUpdated(oldFile: VersionedFile, newContent: string) {
```

Here too -- it's pretty simple self-documenting code only used in one place, so you could just do the check directly in `updateFileContentIfNeeded`.